examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+        # HACK: create a new input point with a given position rather than
+        # copying the existing one. This fixes an issue in some cases where
+        # the points appear to get linked and the @picked_first_ip position
+        # changes when you move the mouse around, even if that should only
+        # affect @mouse_ip.
+        ### @picked_first_ip.copy!(@mouse_ip)
+        @picked_first_ip = Sketchup::InputPoint.new(@mouse_ip.position)
+
+        # Wrap model changes in an operation so it can be undone in one step.
+        # Typically we disable drawing in operations to increase performance,
+        # but here we want live updates when moving the mouse.
+        view.model.start_operation("Move", false)
+
+        # Track whether we are in an operation so we can abort it if the user
+        # activates another tool.
+        @in_operation = true
+
+        # Track the mouse position between frames so we can update the position
+        # of the selected entities.
+        @last_frame_mouse_position = @mouse_ip.position
+      end
+
+      def do_move(view)
+        return unless @last_frame_mouse_position
+
+        # Move entities the same distance input point moved since last frame.
+        frame_move = @mouse_ip.position - @last_frame_mouse_position
+        @last_frame_mouse_position = @mouse_ip.position
+
+        view.model.active_entities.transform_entities(frame_move, @selection)
+      end
+
+      def end_move(view)
+        view.model.commit_operation
+        @in_operation = false
+
+        @picked_first_ip.clear
+        @inference_lock_helper.unlock
+
+        # If tool was activated with a pre-selection, keep the selection and let
+        # the user pick a starting point anywhere. Otherwise, drop the selection
+        # and select whatever the user clicks when starting next move operation.
+        reset_tool unless @preselection
+      end
+
+      def pick_mouse_position(view)
+        if !picked_first_point?
+          @mouse_ip.pick(view, @mouse_position.x, @mouse_position.y)
+        else
+          # If we could, we would want the InputPoint to ignore entities in
+          # @selection. We are currently getting some quite strange behavior
+          # when we get inference from the entities we are already moving.
+          # https://github.com/SketchUp/api-issue-tracker/issues/452
+          @mouse_ip.pick(view, @mouse_position.x, @mouse_position.y, @picked_first_ip)
+          @distance = @mouse_ip.position.distance(@picked_first_ip.position)
+          update_ui
+
+          do_move(view)
+        end
+        view.invalidate
+      end
+
+      def update_ui
+        if @selection.empty?
+          Sketchup.status_text = 'Select a component.'
+        elsif !picked_first_point?
+          Sketchup.status_text = 'Select start point.'
+        else
+          Sketchup.status_text = 'Select end point.'
+
+          # Only update the VCB text if is different from when last set.
```

Comment:
Is this described in the UX guidelines? This sounds interesting, though I'd like to learn more what behaviour this relate to.

Reply:
Clarified the comment. This has to do with key event triggering a UI update, as they could change what position the InputPoints should be picking based on inference locking.

This might be outside the scope of the UX guidlines. It's more about preventing a clearly bugged behavior than a UX design choice.

For the record SketchUp some similar bug where pressing Shift to write certain character on certain keyboards (`<` and `>` I believe) flushes the text the user is writing, making it impossible to enter certain characters.

Reply:
Seems to me the behaviour affects user experience. That if an extension doesn't do this the VCB gets out of sync?

Reply:
There is already a  guideline saying you should use the VCB when applicable. I think that one already implies the ability to write to the VCB.

Reply:
I was referring to what you described as buggy behaviour. If extensions aren't doing what you do here, the VCB values might not represent the state of the tool, right?

Reply:
It would still represent the state of the tool but you wouldn't be able to type anything as the key event would set the VCB value.